libdebugger/break.c

```diff
@@ -156,17 +174,87 @@ list_breakpoints (void)
 
   dbg_msg(  "Num Type           Disp Enb Mask Target  Location");
   for (p = p_breakpoint_top; p; p = p->p_next) {
-    printf("%3u breakpoint     keep   y 0x%02x %s",
-	   p->i_num,
-	   p->brkpt_mask,
-           p->p_target->name);
-    if (p->p_target->floc.filenm) {
-	printf(" at ");
-	print_floc_prefix(&(p->p_target->floc));
+    if (p->type == LINE) {
+      printf("%3u breakpoint     keep   y 0x%02x %s",
+             p->i_num,
+             p->brkpt_mask,
+             p->p_target->name);
+      if (p->p_target->floc.filenm) {
+          printf(" at ");
+          print_floc_prefix(&(p->p_target->floc));
+      }
+    } else {
+      printf("%3u command watchpoint     %s", p->i_num, p->psz_text);
     }
     printf("\n");
   }
 }
+
+bool
+add_command_breakpoint(const char *psz_regex)
+{
+  int re_compile_status;
+  breakpoint_node_t *p_new   = CALLOC (breakpoint_node_t, 1);
+
+  if (!p_new) return false;
+
+  re_compile_status          = regcomp(&p_new->regex, psz_regex, REG_EXTENDED | REG_NOSUB);
+
+  // Not sure if regex_t can be memmoved, so initializing it in the allocated breakpoint_node_t.
+  if (re_compile_status != 0) {
+    char reg_error_message[100];
+    regerror(re_compile_status, &p_new->regex, reg_error_message, sizeof(reg_error_message)-1);
+    dbg_msg("Could not compile regex: %s: %s", psz_regex, reg_error_message);
+    regfree(&p_new->regex);
+    free(p_new);
+    return false;
+  }
+
+  p_new->type                   = COMMAND_WATCH;
+  p_new->psz_text               = strdup(psz_regex);
+  p_new->i_num                  = ++i_breakpoints;
+  p_new->p_target               = NULL;
+  p_new->brkpt_mask             = 0;
+
+  /* Add breakpoint to list of breakpoints. */
+  if (!p_breakpoint_top) {
+    assert(!p_breakpoint_bottom);
+    p_breakpoint_top            = p_breakpoint_bottom = p_new;
+  } else {
+    p_breakpoint_bottom->p_next = p_new;
+  }
+  p_breakpoint_bottom           = p_new;
+
+  return true;
+}
+
+static bool
+any_command_breakpoint_matches(const char *psz_expanded_command)
+{
+  breakpoint_node_t *p;
+
+  for (p = p_breakpoint_top; p; p = p->p_next) {
+    if (p->type == COMMAND_WATCH) {
+      int match_status = regexec(&p->regex, psz_expanded_command, (size_t) 0, NULL, 0);
+      if (match_status == 0) {
+        printf ("Command matched a watchpoint for '%s': %s\n", p->psz_text, psz_expanded_command);
```

Comment:
Following gdb, it might be nice to add *which* breakpoint number matched. 

Also it would be nice to give some idea of *where* the breakpoint matched. To show how this is desireable here is what I tried and what I got. 

```console
$ ./make -X 
remake<0> watch test
remake<1> c
Updating goal targets...
 File 'all' does not exist.
   File 'all-recursive' does not exist.
  Must remake target 'all-recursive'.
Makefile:1403: target 'all-recursive' does not exist
...
##<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
Command matched a watchpoint for 'test': @fail=; \
if (target_option=k; case ${target_option-} in ?) ;; *) echo "am__make_running_with_option: internal error: invalid" "target option '${target_option-}' specified" >&2; exit 1;; esac; has_opt=no; sane_makeflags=$MAKEFLAGS; if { if test -z '0'; then false; elif test -n 'x86_64-pc-linux-gnu'; then true; elif test -n '4.3' && test -n '/tmp/remake'; then true; else false; fi; }; then sane_makeflags=$MFLAGS; else case $MAKEFLAGS in *\\[\ \	]*) bs=\\; sane_makeflags=`printf '%s\n' "$MAKEFLAGS" | sed "s/$bs$bs[$bs $bs	]*//g"`;; esac; fi; skip_next=no; strip_trailopt () { flg=`printf '%s\n' "$flg" | sed "s/$1.*$//"`; }; for flg in $sane_makeflags; do test $skip_next = yes && { skip_next=no; continue; }; case $flg in *=*|--*) continue;; -*I) strip_trailopt 'I'; skip_next=yes;; -*I?*) strip_trailopt 'I';; -*O) strip_trailopt 'O'; skip_next=yes;; -*O?*) strip_trailopt 'O';; -*l) strip_trailopt 'l'; skip_next=yes;; -*l?*) strip_trailopt 'l';; -[dEDm]) skip_next=yes;; -[JT]) skip_next=yes;; esac; case $flg in *$target_option*) has_opt=yes; break;; esac; done; test $has_opt = yes); then \
  failcom='fail=yes'; \
else \
  failcom='exit 1'; \
fi; \
dot_seen=no; \
target=`echo all-recursive | sed s/-recursive//`; \
case "all-recursive" in \
  distclean-* | maintainer-clean-*) list='lib po doc glob unittest libdebugger' ;; \
  *) list='lib po doc glob unittest libdebugger' ;; \
esac; \
for subdir in $list; do \
  echo "Making $target in $subdir"; \
  if test "$subdir" = "."; then \
    dot_seen=yes; \
    local_target="$target-am"; \
  else \
    local_target="$target"; \
  fi; \
  (CDPATH="${ZSH_VERSION+.}:" && cd $subdir && /tmp/remake/./make  $local_target) \
  || eval $failcom; \
done; \
if test "$dot_seen" = "no"; then \
  /tmp/remake/./make  "$target-am" || exit 1; \
fi; test -z "$fail"
wp (/tmp/remake/Makefile:1402)
all-recursive
remake<2>
```

Minimal would be to at least show the offset where the match occurs which available in `regexc`.

Better (and what I do in analogous situtations in other debuggers I've writte) is to show the line containing the match and another line with an `^` to indicate the starting position and match length). For example. Here it would be: 

```
if (target_option=k; case ${target_option-} in ?) ;; *) echo "am__make_running_with_option: internal error: invalid" "target option '${target_option-}' specified" >&2; exit 1;; esac; has_opt=no; sane_makeflags=$MAKEFLAGS; if { if test -z '0'; then false; elif test -n 'x86_64-

                                                                                                                                                                                                                                                                  ^^^^^^		
```

(Except it is a bit of a hassle manually in this format to try to add the `^^^^` in this format.

																		 ```

Reply:
Also, after a breakpoint or watchpoint is hit, the number of times it has hit is updated in `info break`

Reply:
Time to repay karmic debts. When, if not at vacation? Brushing up the PR now.
Let me do it...

Reply:
Warning - I recently broke the 4.3 branch in trying to extend the debugger to allow "info tasks <target-name>". 

That was part of larger scheme to allow `remake --tasks <target-name>` so it would just show the comment for <target-name>. 

If it becomes a problem, I can revert that change and put it in a branch.

Reply:
That's tricky! There are multi-line commands, and there are multi-line matches.
What if I only underline the first line of a match?

Reply:
I added (start position:end position) and an underline that works for a multiline command.
Took me some time, and still there's a room for improvement:
* it doesn't fully underline a multi-line match;
* it should better print the line number of a command matched, if the command is multiline, like in the example.
But I believe it's already workable.